Brokerages/WebSocketClientWrapper.cs

```diff
@@ -0,0 +1,265 @@
+
+        /// <summary>
+        /// Wraps message event
+        /// </summary>
+        public event EventHandler<WebSocketMessage> Message;
+
+        /// <summary>
+        /// Wraps error event
+        /// </summary>
+        public event EventHandler<WebSocketError> Error;
+
+        /// <summary>
+        /// Wraps open method
+        /// </summary>
+        public event EventHandler Open;
+
+        /// <summary>
+        /// Wraps close method
+        /// </summary>
+        public event EventHandler<WebSocketCloseData> Closed;
+
+        /// <summary>
+        /// Event invocator for the <see cref="Message"/> event
+        /// </summary>
+        protected virtual void OnMessage(WebSocketMessage e)
+        {
+            //Logging.Log.Trace("WebSocketWrapper.OnMessage(): " + e.Message);
+            Message?.Invoke(this, e);
+        }
+
+        /// <summary>
+        /// Event invocator for the <see cref="Error"/> event
+        /// </summary>
+        /// <param name="e"></param>
+        protected virtual void OnError(WebSocketError e)
+        {
+            Log.Error(e.Exception, $"WebSocketClientWrapper.OnError(): (IsOpen:{IsOpen}, State:{_client.State}): {e.Message}");
+            Error?.Invoke(this, e);
+        }
+
+        /// <summary>
+        /// Event invocator for the <see cref="Open"/> event
+        /// </summary>
+        protected virtual void OnOpen()
+        {
+            Log.Trace($"WebSocketClientWrapper.OnOpen(): Connection opened (IsOpen:{IsOpen}, State:{_client.State}): {_url}");
+            Open?.Invoke(this, EventArgs.Empty);
+        }
+
+        /// <summary>
+        /// Event invocator for the <see cref="Close"/> event
+        /// </summary>
+        protected virtual void OnClose(WebSocketCloseData e)
+        {
+            Log.Trace($"WebSocketClientWrapper.OnClose(): Connection closed (IsOpen:{IsOpen}, State:{_client.State}): {_url}");
+            Closed?.Invoke(this, e);
+        }
+
+        private async Task HandleConnection(CancellationTokenSource connectionCts)
+        {
+            using (_client = new ClientWebSocket())
+            {
+                Log.Trace("WebSocketClientWrapper.HandleConnection(): Connecting...");
+
+                try
+                {
+                    await _client.ConnectAsync(new Uri(_url), connectionCts.Token);
+                    OnOpen();
+
+                    while ((_client.State == WebSocketState.Open || _client.State == WebSocketState.CloseSent) &&
+                        !connectionCts.IsCancellationRequested)
+                    {
+                        var messageData = await ReceiveMessage(_client, connectionCts.Token);
+
+                        if (messageData.MessageType == WebSocketMessageType.Close)
+                        {
+                            Log.Trace("WebSocketClientWrapper.HandleConnection(): WebSocketMessageType.Close");
+                            return;
```

Comment:
Should we call `OnClose()` here? The wrapping task will try to reconnect, that's expected?

Reply:
> Should we call `OnClose()` here?

This message should be received by the client as a response to the Close request on L105,
OnClose should be called after the close handshake is completed on L120

> The wrapping task will try to reconnect, that's expected?

Correct